src/coord/tests/minimal_resolution.rs

```diff
@@ -0,0 +1,104 @@
+
+use coord::{
+    catalog::builtin::{MZ_CATALOG_SCHEMA, PG_CATALOG_SCHEMA},
+    session::Session,
+};
+use sql::{
+    catalog::Catalog,
+    names::{DatabaseSpecifier, FullName, PartialName},
+};
+
+/// System sessions have an empty `search_path` so it's necessary to
+/// schema-qualify all referenced items.
+#[test]
+fn test_system_session_resolution() -> Result<(), Box<dyn Error>> {
+    let test_cases = vec![
+        (
+            FullName {
+                database: DatabaseSpecifier::Ambient,
+                schema: PG_CATALOG_SCHEMA.to_string(),
+                item: "numeric".to_string(),
+            },
+            PartialName {
+                database: None,
+                schema: Some(PG_CATALOG_SCHEMA.to_string()),
+                item: "numeric".to_string(),
+            },
+        ),
+        (
+            FullName {
+                database: DatabaseSpecifier::Ambient,
+                schema: MZ_CATALOG_SCHEMA.to_string(),
+                item: "mz_array_types".to_string(),
+            },
+            PartialName {
+                database: None,
+                schema: Some(MZ_CATALOG_SCHEMA.to_string()),
+                item: "mz_array_types".to_string(),
+            },
+        ),
+    ];
+
+    let catalog_file = NamedTempFile::new()?;
+    let catalog = coord::catalog::Catalog::open_debug(catalog_file.path())?;
+    let catalog = catalog.for_system_session();
+    for (full_name, partial_name) in test_cases {
+        assert_eq!(catalog.minimal_qualification(&full_name), partial_name);
+    }
+    Ok(())
+}
+
+#[test]
+/// Dummy (and ostensibly client) sessions contain system schemas in their
+/// search paths, so do not require schema qualification on system objects such
+/// as types.
+fn test_dummy_session_resolution() -> Result<(), Box<dyn Error>> {
+    let test_cases = vec![
+        (
+            FullName {
+                database: DatabaseSpecifier::Ambient,
+                schema: PG_CATALOG_SCHEMA.to_string(),
+                item: "numeric".to_string(),
+            },
+            PartialName {
+                database: None,
+                schema: None,
+                item: "numeric".to_string(),
+            },
+        ),
+        (
+            FullName {
+                database: DatabaseSpecifier::Ambient,
+                schema: MZ_CATALOG_SCHEMA.to_string(),
+                item: "mz_array_types".to_string(),
+            },
+            PartialName {
+                database: None,
+                schema: None,
+                item: "mz_array_types".to_string(),
+            },
+        ),
+    ];
+
+    let catalog_file = NamedTempFile::new()?;
+    let catalog = coord::catalog::Catalog::open_debug(catalog_file.path())?;
+    let session = Session::dummy();
+    let catalog = catalog.for_session(&session);
+
+    for (full_name, partial_name) in test_cases {
+        assert_eq!(catalog.minimal_qualification(&full_name), partial_name);
+    }
+    Ok(())
+}
```

Comment:
This is a moment where I think it's nice to double down on the table-driven testing technique to cut down on the amount of shared boilerplate:

```rust
/// System sessions have an empty `search_path` so it's necessary to
/// schema-qualify all referenced items.
///
/// Dummy (and ostensibly client) sessions contain system schemas in their
/// search paths, so do not require schema qualification on system objects such
/// as types.
#[test]
fn test_minimal_qualification() -> Result<(), Box<dyn Error>> {
    struct TestCase {
        input: FullName,
        system_output: PartialName,
        normal_output: PartialName,
    }

    let test_cases = vec![
        TestCase {
            input: FullName {
                database: DatabaseSpecifier::Ambient,
                schema: PG_CATALOG_SCHEMA.to_string(),
                item: "numeric".to_string(),
            },
            system_output: PartialName {
                database: None,
                schema: Some(PG_CATALOG_SCHEMA.to_string()),
                item: "numeric".to_string(),
            },
            normal_output: PartialName {
                database: None,
                schema: None,
                item: "numeric".to_string(),
            },
        },
        TestCase {
            input: FullName {
                database: DatabaseSpecifier::Ambient,
                schema: MZ_CATALOG_SCHEMA.to_string(),
                item: "mz_array_types".to_string(),
            },
            system_output: PartialName {
                database: None,
                schema: Some(MZ_CATALOG_SCHEMA.to_string()),
                item: "mz_array_types".to_string(),
            },
            normal_output: PartialName {
                database: None,
                schema: None,
                item: "mz_array_types".to_string(),
            },
        },
    ];

    let catalog_file = NamedTempFile::new()?;
    let catalog = coord::catalog::Catalog::open_debug(catalog_file.path())?;
    for tc in test_cases {
        assert_eq!(
            catalog
                .for_system_session()
                .minimal_qualification(&tc.input),
            tc.system_output
        );
        assert_eq!(
            catalog
                .for_session(&Session::dummy())
                .minimal_qualification(&tc.input),
            tc.normal_output
        );
    }
    Ok(())
}
```

Reply:
💡 nice, tysm for doing so much of the typing for me